File: .autoloop/programs/tsb-perf-evolve/code/benchmark.ts
// tsb-side benchmark for Series.sortValues.
// Output: a single JSON line on stdout with the shape
//   {"function": "Series.sortValues", "mean_ms": <number>, "iterations": <number>, "total_ms": <number>}
//
// Dataset shape and iteration counts come from ./config.yaml — keep this file
// and ./benchmark.py in lockstep.

import { Series } from "../../../../src/index.ts";

// Inlined from config.yaml — the autoloop agent should keep these in sync.
// (No YAML parser dependency to keep this benchmark hermetic.)
const DATASET_SIZE = 100_000;
const NAN_RATIO = 0.05;
const WARMUP_ITERATIONS = 5;
const MEASURED_ITERATIONS = 50;
const RANDOM_SEED = 42;

// A tiny deterministic PRNG (mulberry32). Note: this is *not* the same
// algorithm as numpy's default_rng on the Python side, so for any given seed
// the two benchmarks will see different concrete values. They will still see
// the same *distribution* (uniform over [-500_000, 500_000) with the same NaN
// fraction), and that is what matters for a sorting micro-benchmark — the
// dataset shape, not the exact bit pattern. If you ever need byte-identical
// inputs across the two sides, swap mulberry32 for a portable PRNG that has a
// matching numpy implementation (e.g. PCG64).
function mulberry32(seed: number): () => number {
  let a = seed >>> 0;
  return () => {
    a = (a + 0x6d2b79f5) >>> 0;
    let t = a;
    t = Math.imul(t ^ (t >>> 15), t | 1);
    t ^= t + Math.imul(t ^ (t >>> 7), t | 61);
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

function buildData(): readonly (number | null)[] {
  const rng = mulberry32(RANDOM_SEED);
  const out: (number | null)[] = new Array(DATASET_SIZE);
  for (let i = 0; i < DATASET_SIZE; i++) {
    out[i] = rng() < NAN_RATIO ? null : rng() * 1_000_000 - 500_000;
  }
  return out;
}

function nowMs(): number {
  return performance.now();
}

function main(): void {
  const data = buildData();
  const series = new Series<number | null>({ data, dtype: "float64" });

  // Warm-up — let the JIT specialize.
  for (let i = 0; i < WARMUP_ITERATIONS; i++) {
    series.sortValues();
  }

  const start = nowMs();
  for (let i = 0; i < MEASURED_ITERATIONS; i++) {
    series.sortValues();
  }
  const totalMs = nowMs() - start;
  const meanMs = totalMs / MEASURED_ITERATIONS;

  const result = {
    function: "Series.sortValues",
    mean_ms: meanMs,
    iterations: MEASURED_ITERATIONS,
    total_ms: totalMs,
  };
  process.stdout.write(`${JSON.stringify(result)}\n`);
}

main();

File: .autoloop/programs/tsb-perf-evolve/code/benchmark.py
"""pandas-side benchmark for Series.sort_values.

Output: a single JSON line on stdout with the shape
    {"function": "Series.sort_values", "mean_ms": <number>,
     "iterations": <number>, "total_ms": <number>}

Dataset shape and iteration counts mirror ./benchmark.ts — keep the two in
lockstep. Fixed seed for reproducibility across runs.
"""

from __future__ import annotations

import json
import sys
import time

import numpy as np
import pandas as pd

# Inlined from config.yaml (kept in sync with benchmark.ts).
DATASET_SIZE = 100_000
NAN_RATIO = 0.05
WARMUP_ITERATIONS = 5
MEASURED_ITERATIONS = 50
RANDOM_SEED = 42


def build_data() -> pd.Series:
    rng = np.random.default_rng(RANDOM_SEED)
    values = rng.uniform(-500_000.0, 500_000.0, size=DATASET_SIZE)
    nan_mask = rng.random(size=DATASET_SIZE) < NAN_RATIO
    values[nan_mask] = np.nan
    return pd.Series(values, dtype="float64")


def main() -> None:
    series = build_data()

    # Warm-up.
    for _ in range(WARMUP_ITERATIONS):
        series.sort_values()

    start = time.perf_counter()
    for _ in range(MEASURED_ITERATIONS):
        series.sort_values()
    total_s = time.perf_counter() - start
    total_ms = total_s * 1000.0
    mean_ms = total_ms / MEASURED_ITERATIONS

    result = {
        "function": "Series.sort_values",
        "mean_ms": mean_ms,
        "iterations": MEASURED_ITERATIONS,
        "total_ms": total_ms,
    }
    sys.stdout.write(json.dumps(result) + "\n")


if __name__ == "__main__":
    main()
